src/ripple/consensus/Validations.h

```diff
@@ -0,0 +1,504 @@
+
+            // Sequence of 0 indicates a missing sequence number
+            if (oldSeq && newSeq && oldSeq == newSeq)
+            {
+                result = AddOutcome::sameSeq;
+
+                // Remove current validation from the ledger set
+                // for the revoked signing key
+                if (val.key() != oldVal.key())
+                {
+                    auto set = byLedger_.find(oldVal.ledgerID());
+                    if (set != byLedger_.end())
+                    {
+                        set->second.erase(key);
+                        // Erase the set if it is now empty
+                        if (set->second.empty())
+                            byLedger_.erase(set);
+                    }
+                }
+            }
+
+            if (val.signTime() > oldVal.signTime() || val.key() != oldVal.key())
+            {
+                // This is either a newer validation or a new signing key
+                auto const oldID = oldVal.ledgerID();
+                // Allow impl to take over oldVal
+                impl().onStale(std::move(oldVal));
+                // Replace old val in the map and set the previous ledger ID
+                ins.first->second = val;
+                ins.first->second.setPreviousLedgerID(oldID);
+            }
+            else
+            {
+                // We already have a newer validation from this source
+                result = AddOutcome::stale;
+            }
+        }
+
+        return result;
+    }
+
+    /** Expire old validation sets
+
+        Remove validation sets that were accessed more than
+        VALIDATION_SET_EXPIRES ago.
+    */
+    void
+    expire()
+    {
+        beast::expire(byLedger_, parms_.VALIDATION_SET_EXPIRES);
+    }
+
+    struct ValidationCounts
+    {
+        //! The number of trusted validations
+        std::size_t count;
+        //! The highest trusting node ID
+        NodeID highNode;
+    };
+
+    /** Distribution of current trusted validations
+
+        Calculates the distribution of current validations but allows
+        ledgers one away from the current ledger to count as the current.
+
+        @param t Current time used to determine ledger staleness
+        @param currentLedger The identifier of the ledger we believe is current
+        @param priorLedger The identifier of our previous current ledger
+        @param cutoffBefore Ignore ledgers with sequence number before this
+        @param j Journal for logging
+    */
+    hash_map<LedgerID, ValidationCounts>
+    currentTrustedDistribution(
+        NetClock::time_point t,
+        LedgerID const& currentLedger,
+        LedgerID const& priorLedger,
+        std::size_t cutoffBefore,
+        beast::Journal& j)
+    {
+        bool valCurrentLedger = currentLedger != beast::zero;
+        bool valPriorLedger = priorLedger != beast::zero;
+
+        hash_map<LedgerID, ValidationCounts> ret;
+
+        current(t, [&](auto const&, auto const& v) {
```

Comment:
I use `auto` params in lambdas for two reasons:

1) When I need the param to be a template
2) When I want a terse syntax

When I don't need a generic lambda, I would apply the same rule of thumb for auto parameters as I do for auto variables. For example, if the lambdas is used near to where it is defined, and the parameter types are easily found at the nearby caller, `auto` is acceptable.

In this case, you can make a strong argument for declaring the parameter types. However, I don't want a rule that says "if a lambda doesn't have to be generic don't use `auto` params".